### PR TITLE
Added Hotkey Settings in the preferences

### PR DESCRIPTION
**Pull request by gtxzsxxk**
Dear maintainers,
Different people have different using habits. Such as the undo. Many people will set them as "Ctrl+Y" rather than "Ctrl+Shift+Z". So I spend some time creating a new tab in the preferences, allowing users to change the hotkeys to whatever they like.
My changes are as follows.

* @gtxzsxxk (2023-07-30)
  * Added a new class HotkeyOptions, a new TAB in the preferences.
  * Modified Menu class, setting up a new abstract method so that the hotkey can be sync via the gui
  * Modified MenuFile MenuEdit MenuSimulate to implement that abstract method
  * After my modification, 11 menuitem in the menu-bar are now available in hotkey settings. Changes will be updated
    immediately.
  * Also, after modified more classes, which can be clearly
    viewed in the FILE CHANGED, I added 26 other hotkey bindings, supporting changing the hotkeys
    for component rotation/direction, auto-labeler and gate-modifier.
  * With more options added, I use JScrollPane to get them all in.
  * And there are some important key bindings that have clear meanings
    and are accepted by nearly each person, I set them as uneditable.
  * However uneditable, people can treat this tab as a 
    key-binding reference.
  * In addition, I'd like to insert more hotkeys to make logisim-evolution more fun.
  * Such as press *SPACE* to flip a splitter or transistor, 
    getting their image along the Y axis.
 

Looking forward to your acceptance.

### CODE REVIEW

File: src/main/resources/resources/logisim/strings/gui/gui.properties [hunk 1]
```diff
@@ -621,6 +621,26 @@ windowGridColorsReset = Reset grid colors to Logisim's defaults
...
+hotkeySimTickFull = Manual Tick Full Cycle
+hotkeySimTickEnabled = Auto-Tick Enabled
+hotkeyEditUndo = Edit Undo
+hotkeyEditRedo = Edit Redo
+hotkeyFileExport = Export File
+hotkeyFilePrint = Print File
+hotkeyFileQuit = Quit Logisim
+hotkeyErrCtrl = This key combo must start with CTRL.
+hotkeyErrConflict = This key is comflicted with :
+#
```
Comment (BFH-ktt1) [inline src/main/resources/resources/logisim/strings/gui/gui.properties:643]: Please add these keys like `# ==> hotkeyOptHeader =` in all the gui_*.properties files. This helps the person who is going to translate which keys are still not available in the given language.

**Review by BFH-ktt1 — CHANGES_REQUESTED** [review]

**Comment by gtxzsxxk** [comment]
Thanks for advising. I'd like to close this and reopen a new PR.

**Comment by maehne** [comment]
@gtxzsxxk: Please, don’t close a PR when changes are necessary due to review comments! Instead, simply do the requested changes in new topical commits, which you push in the same branch of your fork from which you created the PR. Thus, the PR gets automatically updated and the reviewer only gets presented changes, which were added after his previous review. This is especially important if the PR is bigger, i.e., changing/adding/deleting many lines of code in different files.

**Comment by gtxzsxxk** [comment]
I am very sorry for my silly actions..... I promise such things won't happen twice

**Comment by maehne** [comment]
@gtxzsxxk: No problem! We have all been there. It takes a while to get accustomed to Git and the associated Git forge development flows. We appreciate your contribution and just wanted to help you get quickly accustomed to the flow we recommend to use for efficient collaboration.

**Comment by maehne** [comment]
I’ll reopen this PR so that you can address the issues raised by @BFH-ktt1.

**Comment by maehne** [comment]
@gtxzsxxk: That failed. It seems that only the original author can reopen the PR.

**Comment by gtxzsxxk** [comment]
@maehne  thank you for your patience and I think I should withdraw the new PR so that this PR can be opened again

**Review by gtxzsxxk — COMMENTED** [review]
I hope this could work.

File: CHANGES.md [hunk 2]
```diff
@@ -2,32 +2,25 @@
...
-  * Fixed Wrong HDL generation bug in the PortIO component and added the single bit version.
-  * Added TTL 74151: 8-line to 1 line data selector
-  * Added TTL 74153: dual 4-line to 1 line data selector
-  * Added TTL 74181: arithmetic logic unit
-  * Added TTL 74182: look-ahead carry generator
-  * Added TTL 74299: 8-bit universal shift register with three-state outputs
-  * Added TTL 74381: arithmetic logic unit
-  * Added TTL 74541: Octal buffers with three-state outputs
-  * Added TTL 74670: 4-by-4 register file with three-state outputs
+* @gtxzsxxk (2023-07-27)
```
[the remark this answers is not in the document]
Comment (gtxzsxxk) [reply CHANGES.md:5]: done

File: src/main/resources/resources/logisim/strings/gui/gui.properties (hunk 1 above)
Comment (gtxzsxxk) [reply src/main/resources/resources/logisim/strings/gui/gui.properties:643]: done

**Review by davidhutchens — CHANGES_REQUESTED** [review]
I've made some notes about our style and expectations.

But my big issue is that this PR completely breaks the key shortcuts on my Mac. Not everyone is using a computer that uses control- for shortcuts. And they do not work even if I try to set one. It doesn't work (with control or command key) even if I haven't try to change anything. It is broken.

Correction. It actually does work if I use cmd-ctrl-k. But that not good. I expect cmd-k for key k.

Until it works properly on Macs, Windows, and Linux, I cannot support merging this.

**Comment by davidhutchens** [comment]
Having thought about this a bit... The Mac is using exactly the typed character, such as ctrl-k. It also insists on using the cmd key for any shortcut. You machine uses ctrl for commands. If you removed all reference to crtl and allow the user to type the character, such as k, it would work on the Mac. And I strongly suspect that will also work on your machine. I don't think you should insist on it being a control character.

**Comment by gtxzsxxk** [comment]
> Having thought about this a bit... The Mac is using exactly the typed character, such as ctrl-k. It also insists on using the cmd key for any shortcut. You machine uses ctrl for commands. If you removed all reference to crtl and allow the user to type the character, such as k, it would work on the Mac. And I strongly suspect that will also work on your machine. I don't think you should insist on it being a control character.

@davidhutchens thanks for your patience, and I have found out why it would be like this.
In the original code, it is how the hotkeys were bound to the menu-item, take open-a-new-file-item for example: 
`newi.setAccelerator(KeyStroke.getKeyStroke(KeyEvent.VK_N, menuMask));`
You might note that there is a menuMask. It is define here: 
`var menuMask = getToolkit().getMenuShortcutKeyMaskEx();`
The `getMenuShortcutKeyMaskEx` determines which extended modifier key is the appropriate accelerator key for menu shortcuts, and takes `InputEvent.CTRL_DOWN_MASK` as default. However, this method is strongly platform-dependent.

The original code is using menuMask everywhere and I haven't planned to change them. And in my perspective I think the users  might more like to set the keys more explicitly, i.e., the hotkey is what they input. But due to the `menuMask`, I have to ask the users to begin with CTRL. However, this won't work in your mac. 

I wonder whether removing the menuMask and replacing all the Accelerator to be the user's input directly will work. What caused me so hesitated is that it has to change too much original code. I don't know whether it is allowed or good to do so. Let me have a try first.

**Review by gtxzsxxk — COMMENTED** [review]
@davidhutchens @BFH-ktt1 I have cancelled the CTRL-begin limit and committed as you requested. Please check.

**Review by gtxzsxxk — COMMENTED** [review]
I add some new hotkey bindings, allowing users to determine which key to rotate a component and set its direction.

**Comment by gtxzsxxk** [comment]
@maehne @davidhutchens @BFH-ktt1 Excuse me sir, seeing that you haven't replied to my new commits for a while, I wonder if I missed something to check? These days I want to implement a new idea based on my changes, however, if the idea came true while suddenly I found my previous changes are not accepted, my efforts would make no sense.....

**Comment by BFH-ktt1** [comment]
Dear @gtxzsxxk : Do not worry, please take into account that we are in the holiday period, and that we have for our jobs a very tight schedule, sometimes it may take a week for a reaction, or even more. If you find that your PR is not yet in a shape to be merged (read you have more ideas and changes you want to make) please change it to status "draft". The moment you are satisfied you turn it to "review" and you touch only according to the review messages. I hope this helps. I'll be the coming two weeks on holidays, and will probably not react instantly. Do not worry, these kind of efforts are appreciated.
Hope this explains and helps.

**Comment by gtxzsxxk** [comment]
@BFH-ktt1 Thanks for the reply and tips, and be sorry to bother you on your holiday. Got that you’re swamped and have a lot on your plate. I be grateful for your work and help, and I’ll do what you said and update my PR status. Have a nice holiday and talk to you later!

**Comment by davidhutchens** [comment]
To add to @BFH-ktt1's comment. We are volunteers and we have other lives and commitments. Sometimes we respond quickly, sometimes not. It is part of the on-line internet community. We are not even in similar time zones.

I thought I should give some feedback even though you have marked this as draft. So...

I'm surprised. You actually changed the command key to the control key for the Mac. I wouldn't have thought that possible. I also don't really think it is a good idea since no Mac user expects it. But you now have code that allows me to change it back to the command key. That means you are on your way to a solution.

That solution should leave all the keys the way they have been as the default (for all platforms). That way no one is surprised when they open the new version. It should return them to that when they press the button to restore defaults. I think your next task should be to fix the defaults.

**Comment by gtxzsxxk** [comment]
> To add to @BFH-ktt1's comment. We are volunteers and we have other lives and commitments. Sometimes we respond quickly, sometimes not. It is part of the on-line internet community. We are not even in similar time zones.
> 
> I thought I should give some feedback even though you have marked this as draft. So...
> 
> I'm surprised. You actually changed the command key to the control key for the Mac. I wouldn't have thought that possible. I also don't really think it is a good idea since no Mac user expects it. But you now have code that allows me to change it back to the command key. That means you are on your way to a solution.
> 
> That solution should leave all the keys the way they have been as the default (for all platforms). That way no one is surprised when they open the new version. It should return them to that when they press the button to restore defaults. I think your next task should be to fix the defaults.

@davidhutchens I thought my problem was that I don't quite understand Mac's design philosophy. Note that in the current branch, the default modifier to bind with the menuitem is given from
`var menuMask = getToolkit().getMenuShortcutKeyMaskEx();`

In my platform, this mask will be the `CTRL_DOWN_MASK`.
However in mac platform, I guess it will be `command` key or `control` key according to the mac user's preferences, because mac allows user to remap between `command` key and `control` key. I think that is where the problem comes from.

Since We have two solutions.

One solution is to write some platform-dependent codes, which adding some brackets when it is win and when it is mac. When it is mac, we set the mask as `VK_META`, which coresponds to the `command` key. However, what if the mac user expects `control` key?

So I came up with our second solution. It is to use `var menuMask = getToolkit().getMenuShortcutKeyMaskEx();` to reset the default. But the mask in AppPreferences must be initialized with constants. So there must be a lot of coding chores to do.

But it seems that the solution two is more like the original code. I'd like to have a try. Thanks for your comment!

**Comment by davidhutchens** [comment]
This worked in my tests on the Mac. What platform are you developing on? Can someone test it on the third platform?

**Comment by gtxzsxxk** [comment]
> 

@davidhutchens So glad that I saw you comment. I'm using windows, and I luckily have a linux environment. I will head to test it in ubuntu. But I thought we'd better run more tests on other keyboards which differ in layouts.

File: CHANGES.md [hunk 3]
```diff
@@ -2,127 +2,149 @@
...
+        * Redo
+        * Export
+        * Print
+        * Quit
+    * I'd like to insert more hotkeys to make logisim-evolution more fun.
+    * Such as press *SPACE* to rotate the component to be placed on.
+
+
 * @dev (????-??-??)
-  * Added support for scanning 7-segment display on FPGA-boards
```
[the remark this answers is not in the document]
Comment (davidhutchens) [reply CHANGES.md:6]: You marked this conversation as resolved, but you have not fixed the CHANGES.md file as I requested. Your change should be described in one line, placed immediately after the @dev line. That line specifies it is the development version and does not intend to reference any individual.

**Review by davidhutchens — CHANGES_REQUESTED** [review]
You need to merge the current main branch with your work. As presented it would undo many things we have just done.

There is also an issue that the build apparently failed on some tests. I'm not sure why that is since you shouldn't be changing anything related to those tests. But it needs to be resolved before we can merge this.

**Comment by gtxzsxxk** [comment]
> There is also an issue that the build apparently failed on some tests. I'm not sure why that is since you shouldn't be changing anything related to those tests. But it needs to be resolved before we can merge this.

@davidhutchens I don't understand why the build tests would fail at all. I have never touched the test folder at all, but this failure appeared and keeped troubling me since I made my first PR here. 

```
HexFileTest > testSaveLoadMemoryContents(int[]) > [2] [0, 1, 8] FAILED
    java.lang.NoClassDefFoundError at HexFileTest.java:110
        Caused by: java.lang.ExceptionInInitializerError at HeadlessToolkit.java:141
```

This is where he fails. I run the gradle script locally, all passed. I run the gradle script in my wsl, all passed. But when in github actions it occurred errors above. It seems there are some problems in HexFileTest.java. But in my changed files, there is no file related to that. I wonder how the previous PRs prevent this. Could you please give me some hints? And I'll keep digging this too.

**Comment by gtxzsxxk** [comment]
> ```
> HexFileTest > testSaveLoadMemoryContents(int[]) > [2] [0, 1, 8] FAILED
>     java.lang.NoClassDefFoundError at HexFileTest.java:110
>         Caused by: java.lang.ExceptionInInitializerError at HeadlessToolkit.java:141
> ```

Haha, I believe I have found the reason why it would give off such a exception FINALLY! After digging for hours, I realized that the strange problem is coming from `AppPreferences.java`. This is the guilty code:
```
/* Hotkey Settings */
  private static final int hotkeyMenuMask = new JMenu().getToolkit().getMenuShortcutKeyMaskEx(); //Guilty
  public static final PrefMonitor<KeyStroke> HOTKEY_SIM_AUTO_PROPAGATE =
      create(new PrefMonitorKeyStroke("hotkeySimAutoPropagate", KeyEvent.VK_E, hotkeyMenuMask));
```
If we normally run our code, it is ok. But when we run it in github actions, the code is headless, JMenu and his toolkit are not initialized as we expected. Then HeadlessToolkit.java gives a exception, then causes the chain reaction.

So the way to fix is to find the other way to set the hotkeyMenuMask's default value, or simply prevent the headless mode's throwing error. I'm heading it now.

File: src/main/java/com/cburch/logisim/std/gates/GateKeyboardModifier.java [hunk 4]
```diff
@@ -12,75 +12,77 @@
...
-          if (CreateAction) act.set(comp, GateAttributes.ATTR_SIZE, GateAttributes.SIZE_NARROW);
-          else {
-            attrs.setValue(GateAttributes.ATTR_SIZE, GateAttributes.SIZE_NARROW);
-            canvas.repaint();
-          }
+    String compare = InputEvent.getModifiersExText(modifier) + " + " + KeyEvent.getKeyText(KeyCode);
+    /* Here originally have VK_N VK_EQUALS VK_PLUS VK_MINUS but not-used */
+    if (compare.equals(compare.equals(((PrefMonitorKeyStroke) AppPreferences.HOTKEY_GATE_MODIFIER_SIZE_SMALL).getCompareString()))) {
+      if (attrs.containsAttribute(GateAttributes.ATTR_SIZE)) {
+        if (CreateAction) act.set(comp, GateAttributes.ATTR_SIZE, GateAttributes.SIZE_NARROW);
```
Comment (davidhutchens) [inline src/main/java/com/cburch/logisim/std/gates/GateKeyboardModifier.java:35]: Use curly brackets and put the subordinate statement on a different line.

I know that the original was the way it is here. We are working at modifying our code to meet google java coding standards so we are trying to fix things that we touch. The standard is to always use the curly brackets and put subordinate code on a new line. You need not fix the entire file(s), but we would appreciate your doing part you modify.

Do this in other places where you modified or added code.

Your code in new files looked good.

File: src/main/java/com/cburch/logisim/util/AutoLabel.java [hunk 5]
```diff
@@ -213,60 +217,55 @@ public String askAndSetLabel(String componentName, String oldLabel, Circuit circ
...
-          if (createAction)
-            act.set(comp, StdAttr.LABEL_VISIBILITY, false);
-          else
-            attrs.setValue(StdAttr.LABEL_VISIBILITY, false);
-        }
-        return true;
+      return true;
+    } else if (compare.equals(((PrefMonitorKeyStroke) AppPreferences.HOTKEY_AUTO_LABEL_TOGGLE).getCompareString())) {
+      if (attrs.containsAttribute(StdAttr.LABEL_VISIBILITY)) {
+        if (createAction)
```
Comment (davidhutchens) [inline src/main/java/com/cburch/logisim/util/AutoLabel.java:242]: Add curly brackets here and at similar places.

**Review by davidhutchens — COMMENTED** [review]
I am really glad you found the reason the tests were failing! Running headless is an important use case since instructors can run headless with command line arguments to test student work in an automated fashion. It would not be good to break that.

In playing with it, it seemed that some of the single letter actions such as + - w were not working for me. I need to do more to see if that is something you did or something that was preexisting or if I am just misunderstanding what they do. Are they working for you or is it still a work in progress?

**Comment by gtxzsxxk** [comment]
> In playing with it, it seemed that some of the single letter actions such as + - w were not working for me. I need to do more to see if that is something you did or something that was preexisting or if I am just misunderstanding what they do. Are they working for you or is it still a work in progress?

@davidhutchens Thanks for your suggestions every time! I have done all my work and are ready for your reviewing. 

In fact I am still not clear about what your not working means. Let me try explaining it first. Are you seeing that some keys' setting button are disabled? 

LBA%Z~A )9ZN%JQ@FOJ9L$S


Single letter actions such as +, -, W are not working for me too. Because they are not allowed to change due to the reason that they have clear meanings and functions. And not all the hotkeys are involved in the tab, so setting them may cause some potential conflicts.

But reading your words again, I thought I cannot be that lazy. I should involve all the hotkeys, making it a complete system. Heading it now.

**Comment by davidhutchens** [comment]
When I said W didn't work, I meant that trying to use w while dragging a gate (such as AND) onto the canvas would not change the size of the gate as it does in the main branch. That seems to be fixed today.

However, + and - still do not work. I would try with the numeric keypad, but my Mac, like most, does not have a numeric keypad. It is necessary that the standard keys work as before.

I also tried the new U and D keys to move the circuits up and down. That works. However, if you select a circuit and put a text field on it, then type text containing U and D it will move the circuit up or down while you are typing. That is definitely not good. I think it is necessary to think harder about side effects of these non-meta-key hotkeys. Standard keys mean different things in different contexts. When typing text, they should just type text. When dragging they can take on actions. When typing labels they should just type labels. Yes, this version also moves the circuit while typing a label for a component on that circuit.

Or maybe, we should not think of regular characters as hot at all. Careful thinking is needed.

**Comment by gtxzsxxk** [comment]
> However, + and - still do not work. I would try with the numeric keypad, but my Mac, like most, does not have a numeric keypad. It is necessary that the standard keys work as before.

Dear @davidhutchens,, I have done my best to make sure the standard keys work as before, with the least change from the main branch. Do the original code's + and - operations work without a numeric keypad?

> I also tried the new U and D keys to move the circuits up and down. That works. However, if you select a circuit and put a text field on it, then type text containing U and D it will move the circuit up or down while you are typing. That is definitely not good. I think it is necessary to think harder about side effects of these non-meta-key hotkeys. Standard keys mean different things in different contexts. When typing text, they should just type text. When dragging they can take on actions. When typing labels they should just type labels. Yes, this version also moves the circuit while typing a label for a component on that circuit.
> 
> Or maybe, we should not think of regular characters as hot at all. Careful thinking is needed.

Thanks for your reminding, I do be a careless person that time. But that doesn't seems to have a good fix plan in general. Suppose that we are typing some text, then we want to redo the text. However the user let the redo key be  Shift+S. Haha, so every time the user touches a big S by pressing Shift+S into the label, then the text gets redo-ed. It is very funny. I guess the key here is the menuMask. If we press with the menuMask, the system will only take it as a hotkey, instead of the user's text input. Maybe that is why the menuMask be invented by the java designers. Only use the menuMask can we prevent the potential mistakes. What about we force the user to have Ctrl & Cmd key involved when they are setting a new binding as I firstly did, in case of the above situation?  And this will get more complex if we just focus on "When typing labels they should just type labels". The cheapest way might be we directly ban the non-meta-key hotkeys.

Yours sincerely,
Hanyuan

**Comment by gtxzsxxk** [comment]
And if you don't have numeric pad, why not try using the hotkey settings, remapping the plus and minus keys directly to some common keys on the pad, such as Q and Z? When I'm typing Q and Z when an AND gate focused, it works. And when I am typing Q and Z when writing a label, no action the gate does because it is not focused. 

I think what makes the hotkey problem so knotty is that the process of the keyevent are too fragmented. The hotkeys in the menu are very easy to deal with, because they are floating on the surface. But we need to prevent it conflicting with the deeper action keys that hide in the gates and other modules. And different modules have different use of the keys, which makes it harder.

To me I think force the user to add a meta key and ban his modifying the deeper key bindings might be the best anti-conflict way. What do you think of this problem? Looking forward to your review.

**Comment by davidhutchens** [comment]
For your first question, yes, the + and - keys work as expected in the main branch. I didn't try changing the key since that wouldn't keep the original keys working as they did.

I am inclined to agree that requiring a meta-key is the cleanest way to prevent conflicting meanings. The deeper key bindings are scattered in the code so I also agree that the simplest solution is to not allow them to change.

**Comment by gtxzsxxk** [comment]
> For your first question, yes, the + and - keys work as expected in the main branch. I didn't try changing the key since that wouldn't keep the original keys working as they did.

@davidhutchens I checked my codes and found where the bug hid. Now this should be already fixed in my new commits.

> I am inclined to agree that requiring a meta-key is the cleanest way to prevent conflicting meanings. The deeper key bindings are scattered in the code so I also agree that the simplest solution is to not allow them to change.

I have asked all the hotkey bindings of the menu items to start with the meta key. And the remaining key bindings in the preference are very shallow implemented in the code. Thus I directly replace them with my code. This should work without any conflicts now, because the original code itself solved the conflict problem. I just replace them, so there should not have more conflict problems. Haha!

Wish you a good day

**Comment by davidhutchens** [comment]
I have looked at this some today. It looks good. But there is more I want to check.

Have you checked the latest version on Linux as well as Windows?

**Comment by gtxzsxxk** [comment]
> I have looked at this some today. It looks good. But there is more I want to check.
> 
> Have you checked the latest version on Linux as well as Windows?

Sure, I am a windows user, and I also have a ubuntu in my computer. I also test it on mac with my friends' help.

**Review by davidhutchens — CHANGES_REQUESTED** [review]
Overall, this is looking good. I have some minor issues.

The delete key is broken on the Mac when trying to delete objects from the canvas. I marked the place in the code where I think the problem occurs.

I also noted some minor formatting issues with having no space before {.

You are including 3 png files in docs/img that don't seem to be used anywhere. If they are not used, remove them. If they are, please explain where.

**Comment by davidhutchens** [comment]
I've been playing with this some more. In the Preferences panel, the default says the "Duplicate Selected Object" key is "Insert". But the Duplicate Menu item is cmd-D. cmd-D works. I can't set "Duplicate Selected Object" to cmd-D because it says it is used by "Duplicate". If I set "Duplicate Selected Object" to cmd-L, I can use either cmd-D or cmd-L to duplicate my selection.

By the way, I don't have an "Insert" key.

Should there be both "Duplicate" and "Duplicate Selected Object", or are they actually the same action? If a single action, the default value on the Mac should be cmd-D. If separate actions, the "Duplicate Selected Object" action needs a different default because the Mac doesn't have an "Insert" key.

**Comment by gtxzsxxk** [comment]
Well I thought that the second way might be useful so I modified my code as the second way. I'm glad to hear and follow your ideas.

**Comment by davidhutchens** [comment]
Thank you for the explanation.

I took some time to think about the user interface. The reason I was confused about Insert is that I only knew about the menu duplicate. I think it is possible to clear up some of that confusion with the layout of the panel.

I suggest changing the comment above the table from "This Tab allows..." to something like "Menu shortcuts that require the meta-key". Then split the table in two pieces and label the second part with "Context dependent hotkeys that do not require the meta-key" That way the requirement is clear. The reason for the two entries for duplicate is clear, because you put the Insert one in the second table. Then you can name the menu one "Duplicate" and the other "Duplicate selected values" as you did before.

Note that if you can, and wish to, put the symbol for the meta-key on that platform in the header message instead of "the meta-key", that would be reasonable. It is acceptable to me either way.

Does this make sense to you?

**Comment by davidhutchens** [comment]
I would also like to comment on the pop-up when a key is selected for change. When the key conflicts, the message reads "This key is comflicted with : ...". That should be "This key conflicts with...".

I also wonder if it wouldn't be a smoother interface if that pop-up didn't happen and you simply write the empty string into the field when it is clicked on, read the key, then replace it. It would look like editing in place. Only use a pop-up when the entered key conflicts and that pop-up would only have an "OK" button (and put the old value back in the field).

What do you think of this?

**Comment by gtxzsxxk** [comment]
@davidhutchens yes, your ideas are very practical and good. I'm working on that now.

**Comment by gtxzsxxk** [comment]
Dear @davidhutchens 

> I also wonder if it wouldn't be a smoother interface if that pop-up didn't happen and you simply write the empty string into the field when it is clicked on, read the key, then replace it. It would look like editing in place. Only use a pop-up when the entered key conflicts and that pop-up would only have an "OK" button (and put the old value back in the field). Note that clicking outside the field rather than typing a key should cause the old value to reappear.

I have tried my best to do this but I found that the `Note that clicking outside the field rather than typing a key should cause the old value to reappear` is a little hard to implement due to some swing's mechanism. So I tried another method, imitating a microsoft game keymap settings' style, i.e. I designed a new `JHotkeyInput` control which allows the user to input and has two buttons which will only be visible when the user taps the input. To discard the new input, the user could click the cross mark button. To save the new input, the user could click the check mark button. And when this input is unfocused (such as when we click other input control, but clicking the outside won't cause the input unfocused, which made the `clicking outside the field....` hard to implement), it will exit the edit mode, the original value would reappear. My method might be some kind of different from your idea, but I did my best to follow most of that. Here are the images.

We enter the hotkey binding editing mode:
enter_mode

We type the keys into the input:
user_input

**Review by davidhutchens — COMMENTED** [review]
This is better. The code looks good. In playing with the interface I notice two odd behaviors.

If I select a item, say Auto-Tick Enable, and type cmd-K (which is the current and default setting), it complains that it conflicts with itself. Would it be possible to just take that value and not complain?

The other item may be related to your difficulty making these changes or something Swing insists on. If so, just note that. But on first selecting the HotKey Settings panel, the first item is already selected. Would it be possible to have nothing selected? I know that it is possible to have nothing selected, since it is true when a new value is entered. It is also true when Reset to Default is pressed.

**Comment by gtxzsxxk** [comment]
@davidhutchens Could you please give me more information and images about the `I do not see the red X in my interface.` and the second situation? I cannot reproduce those problems...

**Comment by davidhutchens** [comment]
This is what my panel looks like when I first open the panel:

Screenshot 2023-08-14 at 10 22 20 PM

**Comment by gtxzsxxk** [comment]
@davidhutchens Thanks. Let me do some troubleshooting to find some potential bugs to kill. These problems doesn't occur on my win11 and ventura 13.5. I will redesign the layout management and unfocus all controls when they are on load by force and see if the two problems can be avoided.

**Comment by davidhutchens** [comment]
I am running Ventura 13.4.1.

**Comment by gtxzsxxk** [comment]
I don't understand why the first control would be automatically focused and selected...sigh...

**Comment by davidhutchens** [comment]
For me, it looks the same if I compile and run with Java 16.

**Comment by gtxzsxxk** [comment]
Thanks for your info. We have been using nearly the same OS environment. I am running on java 20 too. Wait, are you using another UI theme? I note that our fonts are not the same, and the styles of our borders are not the same too.

**Comment by davidhutchens** [comment]
I have my interface set in preferences to Mac OS X under Look and Feel in the Windows panel.

**Comment by gtxzsxxk** [comment]
Maybe the ui theme is one factor perhaps. But I am really 100% confused why the first item is auto focused & selected. This doesn't occur on my schoolmates' computers and mine..

**Comment by davidhutchens** [comment]
Odd. Maybe I have some strange setting in my Mac preferences. I'll look more tomorrow.

**Comment by gtxzsxxk** [comment]
@davidhutchens catched! I reproduced the two problems. They do come from the window->look and feel. Let me deal with that.

**Review by gtxzsxxk — COMMENTED** [review]
I spend a whole day on the X mark

**Review by davidhutchens — COMMENTED** [review]
You missed some lines in cleanup. But otherwise the code looks good and the interface feels nice.

**Review by gtxzsxxk — COMMENTED** [review]
clear the commented out code

**Review by davidhutchens — APPROVED** [review]
This looks good to me.

**Review by BFH-ktt1 — APPROVED** [review]

**Pull request merged**